File: apps/mobile/app/src/main/kotlin/org/fossify/calendar/continuum/ProposeTimes.kt
package org.fossify.calendar.continuum

import org.joda.time.DateTime

/**
 * Mirror of packages/shared proposeMeetingTimes — timestamps in Unix seconds.
 */
object ProposeTimes {
    fun proposeMeetingTimes(
        busyRanges: List<Pair<Long, Long>>,
        fromTs: Long = DateTime.now().millis / 1000L,
        days: Int = 7,
        durationMinutes: Int = 30,
        count: Int = 5,
        workingHours: ContinuumWorkingHours = ContinuumWorkingHours(),
        travelBufferMinutes: Int = 0,
    ): List<FreeSlot> {
        val durationSec = durationMinutes.coerceAtLeast(1) * 60L
        val buffer = travelBufferMinutes.coerceAtLeast(0) * 60L
        val busy = busyRanges
            .map { (it.first - buffer) to (it.second + buffer) }
            .sortedBy { it.first }
        val proposals = mutableListOf<FreeSlot>()
        var cursor = DateTime(fromTs * 1000L).withTimeAtStartOfDay()

        var d = 0
        while (d < days && proposals.size < count) {
            val day = cursor.plusDays(d)
            val (winStart, winEnd) = dayWindow(day, workingHours)
            var t = maxOf(winStart, fromTs)
            while (t + durationSec <= winEnd && proposals.size < count) {
                val slotEnd = t + durationSec
                val blocker = busy.firstOrNull { t < it.second && slotEnd > it.first }
                if (blocker == null) {
                    proposals.add(FreeSlot(t, slotEnd))
                    t = slotEnd
                } else {
                    t = maxOf(blocker.second, t + 15 * 60L)
                }
            }
            d++
        }
        return proposals
    }

    private fun dayWindow(day: DateTime, hours: ContinuumWorkingHours): Pair<Long, Long> {
        val (sh, sm) = parseHm(hours.start)
        val (eh, em) = parseHm(hours.end)
        val start = day.withTime(sh, sm, 0, 0).millis / 1000L
        val end = day.withTime(eh, em, 0, 0).millis / 1000L
        return start to end
    }

    private fun parseHm(hm: String): Pair<Int, Int> {
        val parts = hm.split(':')
        val h = parts.getOrNull(0)?.toIntOrNull() ?: 9
        val m = parts.getOrNull(1)?.toIntOrNull() ?: 0
        return h.coerceIn(0, 23) to m.coerceIn(0, 59)
    }
}

File: packages/shared/src/proposeTimes.ts
import type { CalendarEvent, FreeSlot } from './events.js'
import type { WorkingHours } from './settings.js'

function parseHm(hm: string): { h: number; m: number } {
  const [h, m] = hm.split(':').map(Number)
  return { h: h || 0, m: m || 0 }
}

function dayWindow(day: Date, hours: WorkingHours): { start: Date; end: Date } {
  const { h: sh, m: sm } = parseHm(hours.start)
  const { h: eh, m: em } = parseHm(hours.end)
  const start = new Date(day)
  start.setHours(sh, sm, 0, 0)
  const end = new Date(day)
  end.setHours(eh, em, 0, 0)
  return { start, end }
}

/** Propose up to `count` free windows of `durationMinutes` across the next `days` days. */
export function proposeMeetingTimes(
  events: CalendarEvent[],
  options: {
    from?: Date
    days?: number
    durationMinutes?: number
    count?: number
    workingHours?: WorkingHours
    travelBufferMinutes?: number
  } = {},
): FreeSlot[] {
  const from = options.from ?? new Date()
  const days = options.days ?? 7
  const duration = (options.durationMinutes ?? 30) * 60_000
  const count = options.count ?? 5
  const hours = options.workingHours ?? { start: '09:00', end: '17:00' }
  const buffer = (options.travelBufferMinutes ?? 0) * 60_000

  const busy = events
    .filter((e) => !e.allDay && e.busy !== false)
    .map((e) => ({
      start: new Date(e.start).getTime() - buffer,
      end: new Date(e.end).getTime() + buffer,
    }))
    .sort((a, b) => a.start - b.start)

  const proposals: FreeSlot[] = []
  const cursor = new Date(from)
  cursor.setHours(0, 0, 0, 0)

  for (let d = 0; d < days && proposals.length < count; d++) {
    const day = new Date(cursor)
    day.setDate(cursor.getDate() + d)
    const { start: winStart, end: winEnd } = dayWindow(day, hours)
    let t = Math.max(winStart.getTime(), from.getTime())
    while (t + duration <= winEnd.getTime() && proposals.length < count) {
      const slotEnd = t + duration
      const overlaps = busy.some((b) => t < b.end && slotEnd > b.start)
      if (!overlaps) {
        proposals.push({ start: new Date(t), end: new Date(slotEnd) })
        t = slotEnd
      } else {
        const blocker = busy.find((b) => t < b.end && slotEnd > b.start)
        t = blocker ? blocker.end : t + 15 * 60_000
      }
    }
  }
  return proposals
}
